### ISSUE TITLE
New lint: `no_relative_paths`

### ISSUE THREAD

**Issue by elenakrittik** [rust-lang/rust-clippy#14605]
### What it does

This allow-by-default lints forbids writing paths using `super`, `self`, and modules declared in the current one directly by name anywhere in the code. In other words, it forces the code to use absolute imports through `crate`. The lint's suggestion should replace the path in question with an absolute one.

### Advantage

The only *objective* advantage of enabling this lint is making `use`s consistent. When you import a bunch of items from around your crate, rust-analyzer will use the shortest path, meaning some items get imported using from the parent using `super`, some get imported using `module_name` if `module_name` was declared in the current module, and some through `crate`. Using `crate` for all imports makes the import "lists" at the top of files more visually consistent.

### Drawbacks

None, as it's allow-by-default/opt-in.

### Example

```rust
mod my_mod;

use external_crate::Object; // external crate => paths always absolute => always good

use super::another_mod::Item; // bad
use crate::a::really::deep::path::SomethingElse; // good already

fn foo() {
    my_mod::helper_fn(); // bad
}
```

Could be written as:

```rust
mod my_mod;

use external_crate::Object;

use crate::some_mod::advanced::another_mod::Item;
use crate::a::really::deep::path::SomethingElse;

fn foo() {
    crate::some_mod::advanced::this_module::my_mod::helper_fn(); // ideally this should, of course, be extracted into a `use`
}
```

**Comment by samueltardieu** [comment]
Out of curiosity, do you have some references of projects using this kind of constraint that you could share?

**Comment by mi2ebi** [comment]
I haven't made anything with *nested* modules but I (and some friends for help) made https://github.com/latkerlo/latkerlo-jvotci/blob/main/rs/src and everything in it only uses `crate`

**Comment by elenakrittik** [comment]
> Out of curiosity, do you have some references of projects using this kind of constraint that you could share?

I haven't yet enforced it across the whole workspace yet because without a clippy lint this has to be done by-hand, but `crates/metal-parser` from https://github.com/metallang/metal exclusively uses `crate` paths and i quite like it. In fact, the desire to automatically enforce this for the rest of the workspace is exactly the driving force behind this lint proposal.

**Comment by elenakrittik** [comment]
Just found out that not all paths in metal-parser are absolute, so while we're at it, here's an example of what the "inconsistency" looks like: https://github.com/metallang/metal/blob/main/crates/metal-parser/src/expr.rs#L3-L8

Most notably, imports from the current crate are mixed up with external ones

**Comment by jasonaowen** [comment]
This seems related to https://github.com/rust-lang/rust-clippy/issues/12796.

**Comment by stepancheg** [comment]
> The only objective advantage of enabling this lint is making uses consistent.

Another advantage is code refactoring: if one import per line is used, sed-assisted refactoring (like move code between crates, when IDE cannot handle it for one reason and another) is much easier when there are no relative imports: rename files, then use `sed` to patch imports.

Similarly, code search, without IDE: grep for qualified type name will show where it is used.

Both are very helpful in large projects.
